### PR TITLE
Fix bug with adding order metadata fails and refund is not possible

### LINKED ISSUE #2656: attach_intent_info_to_order susceptible to errors saving orders
Sometimes WC Pay users contact support because orders are seemingly not refundable. When you try to refund it, there's no actual button:

Screen Capture on 2021-08-04 at 08-12-52

@luizreis and I ran into this in ticket 4091861-zen (which is now solved, only linking here for historical reasons/context). The telltale sign is that the charge is linked in the order notes, but not in the subtitle thing below the order ID:

Markup on 2021-08-04 at 08:11:52

This happens because `attach_intent_info_to_order()` fails/errors during `$order->payment_complete( $intent_id );`, but crucially, it fails _after_ adding the order note. This results in the `_intent_id` meta missing from the postmeta for the order. Link to that line:

https://github.com/Automattic/woocommerce-payments/blob/2.8.0/includes/class-wc-payment-gateway-wcpay.php#L1151

In the case of 4091861-zen, the problem was that Xero and Google Listings & Ads were conflicting and throwing errors during the order save process, but in this current ticket (4184182-zen), there are no fatal error logs to find. However, the same problem exists for their order number 338318:

Markup 2021-08-04 at 08 04 53

Is there any way to fix, detect, or work around this? We have creds for the staging site in 4184182-zen, and I installed a phpMyAdmin plugin so we can look at the db using that.

cc @tpaksu

### PR DESCRIPTION
**Pull request by oaratovskyi**
Fixes #2656 

**Bug background and investigation**

It's not possible to click the _Refund with WooCommerce payments_ button for some orders.

After an initial investigation, it seemed to be due to missing `_intent_id` from the `wp_postmeta` table. It ends up like this one for different reasons, all of those are related to errors in the method below when different plugins are connected:
https://github.com/Automattic/woocommerce-payments/blob/1c1df7bb704b18758312ba57f100366af9564e11/includes/class-wc-payment-gateway-wcpay.php#L1151

I've made an additional investigation and discovered next:
- The code that is responsible for showing the Refund checks if it can refund the order via the gateway.
- In case of WooCommerce Payments the check is implemented here
https://github.com/Automattic/woocommerce-payments/blob/1c1df7bb704b18758312ba57f100366af9564e11/includes/class-wc-payment-gateway-wcpay.php#L1244

Eventually, it means that the not only `_intent_id` meta item missing is the reason for the bug, but also `_charge_id`. 

**How to reproduce the bug**

I was able to reproduce the missing button on my local after deleting the meta value by key `_charge_id` in the `wp_postmeta` table.

#### Changes proposed in this Pull Request

One of the solutions that were discussed in the slack channel was to change the order of how the code blocks are executed in `WC_Payment_Gateway_WCPay::attach_intent_info_to_order()`.
It might not provide 100% coverage for this bug, but given that covering 100% of the cases here might be virtually impossible it might be good enough. 
Another one was to `try/catch` the errors from the `payment_complete` method. If some 3rd party plugin will throw an error, the code will be executed despite the failure.
I believe that a combination of those two changes will solve the issue for new orders and meta items will be stored properly and then the button would be available.

**Important**: old orders won't be fixes as they are already missing the meta value by key `_charge_id` in the `wp_postmeta` table.

#### Testing instructions

1. Go to the shop as a shopper, add something to the cart, do a checkout, and complete the order.
2. Check that both `_intent_id` and `_charge_id` meta values are present in `wp_postmeta` table.
3. Go to admin as a merchant, find the order and try to refund it. The button should appear. 
4. Do the refund, check that everything works properly.
5. Enable manual charge capture in the WCPay settings. Repeat steps 1-4.
* Unit tests should pass.

-------------------

- [x] Added changelog entry (or does not apply)
- [x] Covered with tests (or have a good reason not to test in description ☝️)
- [x] Tested on mobile (or does not apply)

**Post merge**

- [ ] Added testing instructions to the [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) (or does not apply)

### CODE REVIEW

File: changelog.txt [hunk 1]
```diff
@@ -7,6 +7,7 @@
 * Fix - Fixed missing file error for removed CSS file.
 * Add - Currency deletion confirmation modal for currencies that are bound to an UPE method.
 * Fix - Currency switcher does not affect order confirmation screen prices.
+* Fix - Refund button is missing from orders that saved with errors.
```
Comment (tpaksu) [inline changelog.txt:10]: Will it be better to rephrase it to describe what is done than what the error was? I mean something like 'Fill missing order_intent_info even if an exception occurs`?

**Review by tpaksu — APPROVED** [review]
I was able to reproduce the error setting `_charge_id` column data to null, your code will prevent the field stay empty in the future by previously saving the fields into database before any exception happens, thanks for solving this! LGTM, after solving the one small nitpick I left as a comment, you can :ship: it!

File: changelog.txt (hunk 1 above)
Comment (oaratovskyi) [reply changelog.txt]: Yep, sounds good!
Fixed in https://github.com/Automattic/woocommerce-payments/pull/2936/commits/f567b0db0cbbf0bc4a656fd7645ff165a89ee88a and https://github.com/Automattic/woocommerce-payments/pull/2936/commits/c992f1fa3169e9b4beffc0e21acd67c71701a35c.
